## NGS data analysis for CNV detection 

Starting from a fastq files (1-9) or from BAM files (4-9):

1. Sequence trimming

   Removing low quality bases 

2. Genome alignment

   Locate individual reads to the reference genome

3. Alignment refinement

   Realign around INDELs, remove duplicates, etc...

4. Bin creation and count

   The choice of the bin size has a major importance (moving or overlaping windows)

5. Bin count correction (optional)
  1. GC content
  2. Mappability

   **It won't be done today: it needs a whole genome data set**

6. Computing LRR

   Normalize count between tumor and germaline then estimate the logRatio in each bin

7. LRR signal smoothing (optional)

   Reduce single point outliers impact before analysis

8. LRR signal segmentation

   Copy number aberrations (CNA) occur in contiguous regions of the chromosome that often cover multiple bins up to whole chromosome arms or chromosomes. The segmentation split the chromosomes into regions of equal copy number that accounts for the noise in the data

   Non-exhaustive available methods: 
    * Circular Binary Segmentation
    * Mean Shift-Based
    * Shifting Level Model
    * Expectation Maximization
    * Hidden Markov Model
    * Etc...
9. CNV calling from segments

   use any method that can differenciate the copy number of each segments 

   Non-exhaustive available methods: 
    * Threshold
    * Sd deviation
    * Poisson distribution
    * Z-score
    * Event-Wise Testing
    * Etc...