### bincounter parameters
| Parameter | explanation | Sugested value  |
| ------------- |:-------------:| -----:|
| --minMapQ | Filter out reads a mapping quality lower or equal at  | 35 |
| --gc      | Compute GC% per window |  _not neccessary_ |
| --ref | Path to the indexed Reference Genome |  _not neccessary_ |
| --refbam | Path to the germline sample bam | ../alignment/normal/normal.sorted.dup.bam |
| --bam | Path to the tumor sample bam | ../alignment/tumor/tumor.sorted.dup.bam |
| --norm | unit for count normalization | chr |
| --windows | bin size | 1000 | 


### bincoutner analysis
Feel free to test different sets of parameters.

Sugested command line

```
cd NGS_CNV
java -jar $BVATOOLS_JAR bincounter --minMapQ 35 --refbam ../alignment/normal/normal.sorted.dup.bam --bam ../alignment/tumor/tumor.sorted.dup.bam --norm chr --windows 1000 > sample_binCount_1kb.tsv
```

The output file should looks like:
```
$ head sample_binCount_1kb.tsv 
chr	start	end	sample_raw	ref_raw	sample_normalized	ref_normalized	ln(sample/ref)
chr1	0	199	0	0	0.0	0.0	NaN
chr1	200	399	0	0	0.0	0.0	NaN
chr1	400	599	0	0	0.0	0.0	NaN
chr1	600	799	0	0	0.0	0.0	NaN
chr1	800	999	0	0	0.0	0.0	NaN
chr1	1000	1199	0	0	0.0	0.0	NaN
chr1	1200	1399	0	0	0.0	0.0	NaN
chr1	1400	1599	0	0	0.0	0.0	NaN
chr1	1600	1799	0	0	0.0	0.0	NaN
```
