## NGS data analysis for CNV detection 

Starting from a CEL files (1-9) or from LRR and BAF files (3-9):

####1. Normalize probe intensity

   Normalise sample probe intensity against some reference HapMap pooled samples in order to generate a LRR values 

####2. Generate genotyping from CEL files

   Contrast the normalized signal intensity for each allelic probe of a same SNP in order to generate BAF values

####3. Probe filtering

   Remove homozygote SNPs

####4. simultaneous segmenetation of LRR and BAF signal

   Models of segmentation should fit between the 2 signal


####5. Estimation of the model paremters
  1. aberrant cell fraction
  2. tumor ploidy
  3. absolute allele-specific copy number calls (for each allelic probes of the SNP)
  

####6. CNV calling from segments

   determine the copy number by simply counting the total number of allele reported to the sample ploidy 

[step 3 to 6  - analysis](../analysis/ASCAT.md)